File: Tests/ConnectorControlCoreTests/ProfileTests.swift
import XCTest
@testable import ConnectorControlCore

final class ProfileTests: XCTestCase {
    private func entry(_ url: String) -> MCPEntry {
        MCPEntry(config: RemotePattern.make(url: url))
    }

    // MARK: - Decoding

    func testV2RoundTripPreservesTwoProfiles() throws {
        let store = MasterStore(
            activeProfile: "Work",
            profiles: [
                "Work": Profile(mcps: ["a": entry("https://a.example/mcp")]),
                "Personal": Profile(mcps: ["b": entry("https://b.example/mcp")]),
            ])
        let data = try JSONEncoder().encode(store)
        let decoded = try JSONDecoder().decode(MasterStore.self, from: data)
        XCTAssertEqual(decoded, store)
        XCTAssertEqual(decoded.activeProfile, "Work")
        XCTAssertEqual(decoded.profiles["Work"]?.mcps.keys.sorted(), ["a"])
        XCTAssertEqual(decoded.profiles["Personal"]?.mcps.keys.sorted(), ["b"])
    }

    // MARK: - mcps accessor scoping

    func testMcpsAccessorReadsAndWritesOnlyActiveProfile() {
        var store = MasterStore(
            activeProfile: "Work",
            profiles: [
                "Work": Profile(mcps: ["a": entry("https://a.example/mcp")]),
                "Personal": Profile(mcps: ["b": entry("https://b.example/mcp")]),
            ])
        store.mcps["c"] = entry("https://c.example/mcp")
        XCTAssertEqual(store.profiles["Work"]?.mcps.keys.sorted(), ["a", "c"])
        XCTAssertEqual(store.profiles["Personal"]?.mcps.keys.sorted(), ["b"], "untouched")
    }

    // MARK: - Profile management

    func testAddProfileCopyingCurrent() {
        var store = MasterStore.single(["a": entry("https://a.example/mcp")])
        let error = store.addProfile(named: "Copy", copyingCurrent: true)
        XCTAssertNil(error)
        XCTAssertEqual(store.activeProfile, "Copy")
        XCTAssertEqual(store.profiles["Copy"]?.mcps.keys.sorted(), ["a"])
        XCTAssertEqual(store.profiles["Default"]?.mcps.keys.sorted(), ["a"], "original untouched")
    }

    func testAddProfileEmptyStartsBlank() {
        var store = MasterStore.single(["a": entry("https://a.example/mcp")])
        let error = store.addProfile(named: "Fresh", copyingCurrent: false)
        XCTAssertNil(error)
        XCTAssertEqual(store.profiles["Fresh"]?.mcps, [:])
    }

    func testAddProfileRejectsEmptyName() {
        var store = MasterStore.empty
        XCTAssertEqual(store.addProfile(named: "   ", copyingCurrent: false), "Name must not be empty.")
    }

    func testAddProfileRejectsDuplicateName() {
        var store = MasterStore.empty
        XCTAssertEqual(store.addProfile(named: "Default", copyingCurrent: false),
                       "A profile named “Default” already exists.")
    }

    func testRenameActiveProfile() {
        var store = MasterStore.empty
        let error = store.renameActiveProfile(to: "Main")
        XCTAssertNil(error)
        XCTAssertEqual(store.activeProfile, "Main")
        XCTAssertEqual(Array(store.profiles.keys), ["Main"])
    }

    func testRenameActiveProfileRejectsCollision() {
        var store = MasterStore(
            activeProfile: "Work",
            profiles: ["Work": Profile(), "Personal": Profile()])
        XCTAssertEqual(store.renameActiveProfile(to: "Personal"), "A profile named “Personal” already exists.")
        XCTAssertEqual(store.activeProfile, "Work", "unchanged on error")
    }

    func testRenameActiveProfileRejectsEmptyName() {
        var store = MasterStore.empty
        XCTAssertNotNil(store.renameActiveProfile(to: "  "))
    }

    func testDeleteActiveProfileSwitchesToFirstRemaining() {
        var store = MasterStore(
            activeProfile: "Work",
            profiles: ["Work": Profile(), "Alpha": Profile(), "Zeta": Profile()])
        let error = store.deleteActiveProfile()
        XCTAssertNil(error)
        XCTAssertEqual(store.activeProfile, "Alpha")
        XCTAssertNil(store.profiles["Work"])
    }

    func testDeleteActiveProfileRejectsLastProfile() {
        var store = MasterStore.empty
        XCTAssertEqual(store.deleteActiveProfile(), "Can’t delete the last profile.")
        XCTAssertEqual(store.profiles.count, 1)
    }

    func testSwitchProfile() {
        var store = MasterStore(
            activeProfile: "Work",
            profiles: ["Work": Profile(), "Personal": Profile()])
        XCTAssertNil(store.switchProfile(to: "Personal"))
        XCTAssertEqual(store.activeProfile, "Personal")
    }

    func testSwitchProfileRejectsUnknownName() {
        var store = MasterStore.empty
        XCTAssertEqual(store.switchProfile(to: "Nope"), "No profile named “Nope”.")
        XCTAssertEqual(store.activeProfile, "Default")
    }

    /// windows/tests/ConnectorControl.Core.Tests/ProfileTests.cs
    /// ErrorMessagesUseTypographicPunctuationLikeTheMacApp.
    func testErrorMessagesUseTypographicPunctuationLikeTheMacApp() {
        var store = MasterStore.empty
        let duplicate = store.addProfile(named: "Default", copyingCurrent: false)!
        XCTAssertEqual(duplicate[duplicate.index(before: duplicate.range(of: "Default")!.lowerBound)], "“")
        XCTAssertEqual(duplicate[duplicate.range(of: "Default")!.upperBound], "”")
        XCTAssertTrue(store.deleteActiveProfile()!.contains("’"))
    }
}

File: windows/tests/ConnectorControl.Core.Tests/ProfileTests.cs
using ConnectorControl.Core.Tests.TestSupport;

namespace ConnectorControl.Core.Tests;

public class ProfileTests : IDisposable
{
    private readonly TempDir dir = new("profile-store");
    private string Url => dir.File("mcps.json");

    public void Dispose() => dir.Dispose();

    private static McpEntry Entry(string url) => new(JsonValue.Object(
        ("command", JsonValue.String("npx")),
        ("args", JsonValue.Array([JsonValue.String("-y"), JsonValue.String("mcp-remote"), JsonValue.String(url)]))));

    private static string[] Keys(Profile? p) => p is null ? [] : p.Mcps.Keys.Order(StringComparer.Ordinal).ToArray();

    private static MasterStore TwoProfiles() => new(2, "Work",
    [
        new("Work", new Profile(new Dictionary<string, McpEntry> { ["a"] = Entry("https://a.example/mcp") })),
        new("Personal", new Profile(new Dictionary<string, McpEntry> { ["b"] = Entry("https://b.example/mcp") })),
    ]);

    // Decoding

    [Fact]
    public void V2RoundTripPreservesTwoProfiles()
    {
        var store = TwoProfiles();
        var decoded = MasterStore.FromJson(JsonValue.Parse(store.ToJson().Serialize()));
        Assert.Equal(store, decoded);
        Assert.Equal("Work", decoded.ActiveProfile);
        Assert.Equal(["a"], Keys(decoded.Profiles["Work"]));
        Assert.Equal(["b"], Keys(decoded.Profiles["Personal"]));
    }

    [Fact]
    public void UnknownActiveProfileFallsBackToExistingProfile()
    {
        File.WriteAllText(Url, "{\"version\":2,\"activeProfile\":\"Ghost\",\"profiles\":{\"Alpha\":{\"mcps\":{}},\"Beta\":{\"mcps\":{}}}}");
        var (store, corrupt) = MasterStoreIO.Load(Url);
        Assert.Null(corrupt);
        Assert.Equal("Alpha", store.ActiveProfile);
    }

    [Fact]
    public void V1FormatFileIsTreatedAsCorruptAndRebuilt()
    {
        File.WriteAllText(Url, "{\"version\":1,\"mcps\":{\"scoutbook\":{\"enabled\":true,\"config\":{\"command\":\"npx\",\"args\":[\"-y\",\"mcp-remote\",\"https://example.com/mcp\"]},\"lastEditView\":\"form\"}}}");
        var (store, corrupt) = MasterStoreIO.Load(Url);
        Assert.Equal(MasterStore.Empty(), store);
        Assert.NotNull(corrupt);
        Assert.StartsWith("mcps.corrupt.", Path.GetFileName(corrupt), StringComparison.Ordinal);
        Assert.False(File.Exists(Url));
    }

    [Theory]
    [InlineData("{\"version\":2,\"activeProfile\":\"D\",\"profiles\":{\"D\":{\"mcps\":{\"s\":{\"enabled\":true,\"config\":{}}}}}}")]                    // lastEditView missing
    [InlineData("{\"version\":2,\"activeProfile\":\"D\",\"profiles\":{\"D\":{\"mcps\":{\"s\":{\"enabled\":\"yes\",\"config\":{},\"lastEditView\":\"form\"}}}}}")]  // enabled not bool
    [InlineData("{\"version\":2,\"activeProfile\":\"D\",\"profiles\":{\"D\":{\"mcps\":{\"s\":{\"enabled\":true,\"config\":{},\"lastEditView\":\"grid\"}}}}}")]   // unknown view
    [InlineData("{\"version\":\"2\",\"activeProfile\":\"D\",\"profiles\":{\"D\":{\"mcps\":{}}}}")]                                                  // version not a number
    [InlineData("{\"version\":2,\"profiles\":{\"D\":{\"mcps\":{}}}}")]                                                                             // activeProfile missing
    [InlineData("[]")]
    public void RequiredKeysAreStrictLikeSwiftCodable(string json)
    {
        Assert.Throws<FormatException>(() => MasterStore.FromJson(JsonValue.Parse(json)));
    }

    [Fact]
    public void UnknownKeysAreIgnored()
    {
        var store = MasterStore.FromJson(JsonValue.Parse("{\"version\":2,\"activeProfile\":\"D\",\"future\":1,\"profiles\":{\"D\":{\"mcps\":{},\"note\":\"x\"}}}"));
        Assert.Equal("D", store.ActiveProfile);
    }

    // mcps accessor scoping

    [Fact]
    public void McpsAccessorReadsAndWritesOnlyActiveProfile()
    {
        var store = TwoProfiles();
        store.Mcps["c"] = Entry("https://c.example/mcp");
        Assert.Equal(["a", "c"], Keys(store.Profiles["Work"]));
        Assert.Equal(["b"], Keys(store.Profiles["Personal"]));
    }

    // Profile management

    [Fact]
    public void AddProfileCopyingCurrent()
    {
        var store = new MasterStore(new Dictionary<string, McpEntry> { ["a"] = Entry("https://a.example/mcp") });
        Assert.Null(store.AddProfile("Copy", copyingCurrent: true));
        Assert.Equal("Copy", store.ActiveProfile);
        Assert.Equal(["a"], Keys(store.Profiles["Copy"]));
        Assert.Equal(["a"], Keys(store.Profiles["Default"]));
    }

    [Fact]
    public void AddProfileEmptyStartsBlank()
    {
        var store = new MasterStore(new Dictionary<string, McpEntry> { ["a"] = Entry("https://a.example/mcp") });
        Assert.Null(store.AddProfile("Fresh", copyingCurrent: false));
        Assert.Empty(store.Profiles["Fresh"].Mcps);
    }

    [Fact]
    public void AddProfileRejectsEmptyName()
    {
        Assert.Equal("Name must not be empty.", MasterStore.Empty().AddProfile("   ", false));
    }

    [Fact]
    public void AddProfileRejectsDuplicateName()
    {
        Assert.Equal("A profile named \"Default\" already exists.", MasterStore.Empty().AddProfile("Default", false));
    }

    [Fact]
    public void RenameActiveProfile()
    {
        var store = MasterStore.Empty();
        Assert.Null(store.RenameActiveProfile("Main"));
        Assert.Equal("Main", store.ActiveProfile);
        Assert.Equal(["Main"], store.Profiles.Keys.ToArray());
    }

    [Fact]
    public void RenameActiveProfileRejectsCollision()
    {
        var store = new MasterStore(2, "Work", [new("Work", new Profile()), new("Personal", new Profile())]);
        Assert.Equal("A profile named \"Personal\" already exists.", store.RenameActiveProfile("Personal"));
        Assert.Equal("Work", store.ActiveProfile);
    }

    [Fact]
    public void RenameActiveProfileRejectsEmptyName()
    {
        Assert.NotNull(MasterStore.Empty().RenameActiveProfile("  "));
    }

    [Fact]
    public void DeleteActiveProfileSwitchesToFirstRemaining()
    {
        var store = new MasterStore(2, "Work", [new("Work", new Profile()), new("Alpha", new Profile()), new("Zeta", new Profile())]);
        Assert.Null(store.DeleteActiveProfile());
        Assert.Equal("Alpha", store.ActiveProfile);
        Assert.False(store.Profiles.ContainsKey("Work"));
    }

    [Fact]
    public void DeleteActiveProfileRejectsLastProfile()
    {
        var store = MasterStore.Empty();
        Assert.Equal("Can't delete the last profile.", store.DeleteActiveProfile());
        Assert.Single(store.Profiles);
    }

    [Fact]
    public void SwitchProfile()
    {
        var store = new MasterStore(2, "Work", [new("Work", new Profile()), new("Personal", new Profile())]);
        Assert.Null(store.SwitchProfile("Personal"));
        Assert.Equal("Personal", store.ActiveProfile);
    }

    [Fact]
    public void SwitchProfileRejectsUnknownName()
    {
        var store = MasterStore.Empty();
        Assert.Equal("No profile named \"Nope\".", store.SwitchProfile("Nope"));
        Assert.Equal("Default", store.ActiveProfile);
    }

    [Fact]
    public void CloneIsDeepAndEqual()
    {
        var store = TwoProfiles();
        var clone = store.Clone();
        Assert.Equal(store, clone);
        clone.Mcps["z"] = Entry("https://z.example/mcp");
        Assert.NotEqual(store, clone);
        Assert.False(store.Mcps.ContainsKey("z"));
    }
}
